### PR TITLE
Stabilize training with sparse depth

### PR DESCRIPTION
**Pull request by wangyida**
I found that training with all sparse point estimated with MVS method such as COLMAP may harm the model, while some points are with large error
- [x] filter out depth supervision coming from remote points 
- [x] add a points-dropping ratio in the parser

### CODE REVIEW

File: nerf/renderer.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 import raymarching
 import nvdiffrast.torch as dr
+glctx = dr.RasterizeGLContext(output_db=False)
```
Comment (ashawkey) [inline nerf/renderer.py:16]: why is this needed?

File: nerf/colmap_provider.py [hunk 2]
```diff
@@ -270,13 +270,17 @@ def __init__(self, opt, device, type='train', n_test=24):
                 self.cam_near_far.append([np.min(depth), np.max(depth)])
 
                 # sparse depth info
+                # drop out unconfident and remote points
+                pts_ids = depth < np.min(depth) + (np.max(depth) - np.min(depth))*0.2
+                pts_ids = np.logical_and(weight.astype(np.float32) > 2 - 2*self.opt.sparse_depth_ratio, pts_ids)
```
Comment (ashawkey) [inline nerf/colmap_provider.py:275]: Is there any insight behind this ratio? I think weight already describes the possible estimation error in supervision.

**Comment by ashawkey** [comment]
@wangyida Hi, thanks for your contribution! 
I observed similar thing and haven't tested a lot on sparse depth supervision (it seems to work only in few-shot settings, like <= 10 images). Do you observe better performance with these improvements? Also, what datasets are you using?
There are some small questions I left as comment and hope you could help.

File: nerf/renderer.py (hunk 1 above)
Comment (wangyida) [reply nerf/renderer.py:16]: Found that declaring gl context will crash without defining right after importing dr when I'm using command line mode. Seems that it need to be defined on top of all other classes and object

File: nerf/colmap_provider.py (hunk 2 above)
Comment (wangyida) [reply nerf/colmap_provider.py:275]: found that training with some remote points will harm the convergence of the model when their error is large. The mentioned ratio means the percentage of the depth supervision we shall use as supervision, assuming that the weights are presented in a uniform distribution, although it is a naive assumption. Please inform me in case you have better ideas

**Comment by wangyida** [comment]
> 

Thanks @ashawkey my tests are validated on unbounded scenes including Mipnerf360 datasets including categories of room and garden. PSNR will increase in range of about 0.5 to 1. I agree with your opinion towards the benefit of training with sparse depth. 
It feels like

1. Training with sufficient amount of views will not be obviously improved supervised with depth, sometimes the results will be even worse considering the fact that depth is usually estimated with errors
2. I found that you have adopted a smart way of training with depth which is randomly picking up few samples with depth among all camera views during training [here](https://github.com/ashawkey/nerf2mesh/blob/main/nerf/colmap_provider.py#L510-L511)
3. An error occurs [here](https://github.com/ashawkey/nerf2mesh/blob/main/nerf/colmap_provider.py#L550), informing that rays['j'] not defined, still do not know the reason.
Best

File: nerf/renderer.py (hunk 1 above)
Comment (ashawkey) [reply nerf/renderer.py:16]: I think it may be related to environmental issues on your side since I don't meet this problem. Also, `glctx` is never used (we use `self.glctx`)

**Comment by ashawkey** [comment]
Could you reproduce the error in 3? I can check it too.

**Comment by wangyida** [comment]
> Could you reproduce the error in 3? I can check it too.

One issue is shown as below:
```bash
[INFO] extracting sparse depth info...
100%|███████████████████████████████████████████████████████████| 599/599 [00:00<00:00, 750.69it/s][INFO] extracted 4898.73 valid sparse depth on average per image
Loading val data: 100%|████████████████████████████████████████████| 12/12 [00:00<00:00, 62.76it/s][mark untrained grid] 6875803 from 16777216
==> Start Training Epoch 1, lr=0.000100 ...
  0% 0/300 [00:00<?, ?it/s]Traceback (most recent call last):
  File "main.py", line 244, in <module>
    trainer.train(train_loader, valid_loader, max_epoch)
  File "/home/yidaw/Documents/buildboat/nerf2mesh/nerf/utils.py", line 931, in train
    self.train_one_epoch(train_loader)
  File "/home/yidaw/Documents/buildboat/nerf2mesh/nerf/utils.py", line 1164, in train_one_epoch
    preds, truths, loss_net = self.train_step(data)
  File "/home/yidaw/Documents/buildboat/nerf2mesh/nerf/utils.py", line 796, in train_step
    self.opt.num_rays = int(round((self.opt.num_points / outputs['num_points']) * self.opt.num_rays))
ZeroDivisionError: division by zero
  0% 0/300 [00:00<?, ?it/s]
```
```outputs['num_points']``` is 0 there although sparse points are already loaded for each sample.
have you encoutered such issue?

**Comment by roshanpuru** [comment]
@wangyida Did you downscale the images by any chance before training ?

**Comment by wangyida** [comment]
> downscale

yes, downscaled by factor of 4

**Comment by ashawkey** [comment]
@wangyida I guess this is because there is no valid sparse depth pixels due to the masking? Maybe we need to add an exception handling if `pts_ids` is all False.

**Comment by wangyida** [comment]
There is another memory issue when I'm supervising with the sparse depth, seems  that the encoding for certain rays are out of the boundings
```bash
==> Start Training Epoch 2, lr=0.009852 ...
loss=0.122079 (0.013412), lr=0.009806: :  31% 60/194 [00:00<00:02, 61.26it/s]Traceback (most rece
nt call last):
  File "main.py", line 224, in <module>                                                              trainer.train(train_loader, valid_loader, max_epoch)
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/utils.py", line 927, in train
    self.train_one_epoch(train_loader)                                                             File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/utils.py", line 1161, in train_one_epoch
    preds, truths, loss_net = self.train_step(data)
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/utils.py", line 668, in train_step
    outputs = self.model.render(rays_o, rays_d, index=index, staged=False, bg_color=bg_color, perturb=True, cam_near_far=cam_near_far, shading=shading, **vars(self.opt))
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/renderer.py", line 638, in render
    sigmas, rgbs, speculars = self(xyzs, dirs, ind_code, shading)
  File "/home/yidaw/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/network.py", line 57, in forward
    color, specular = self.rgb(x, d, c, shading)
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/network.py", line 89, in rgb
    geo_feat = self.geo_feat(x, c)
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/network.py", line 80, in geo_feat
    h = self.color_net(h)
  File "/home/yidaw/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/yidaw/Documents/gitfarm/nerf2mesh/nerf/network.py", line 26, in forward
    x = self.net[l](x)
  File "/home/yidaw/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/yidaw/anaconda3/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 103,
in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: CUDA out of memory. Tried to allocate 5.26 GiB (GPU 0; 23.70 GiB total capacity; 15.47 GiB already allocated; 3.99 GiB free; 17.70 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
loss=0.122079 (0.013412), lr=0.009806: :  31% 61/194 [00:01<00:03, 39.55it/s]
```
